pkg/controllers/machine_controller.go

```diff
@@ -268,35 +265,29 @@ func (r *MachineReconciler) startCheckAndEnqueueVolumeResize(ctx context.Context
 }
 
 func (r *MachineReconciler) startCheckAndEnqueueMachineState(ctx context.Context, log logr.Logger) {
-	wait.UntilWithContext(ctx, func(ctx context.Context) {
-		machines, err := r.machines.List(ctx)
-		if err != nil {
-			log.Error(err, "failed to list machines")
-			return
-		}
+	eventChan, err := r.libvirt.LifecycleEvents(ctx)
+	if err != nil {
+		log.Error(err, "failed to subscribe lifecycle events")
+		return
+	}
 
-		for _, machine := range machines {
-			if machine.DeletedAt != nil || !slices.Contains(machine.Finalizers, MachineFinalizer) {
-				continue
-			}
-			id := machine.ID
+	log.Info("subscribing lifecycle events")
 
-			state, err := r.getMachineState(id)
-			if err != nil {
-				if libvirt.IsNotFound(err) {
-					log.V(1).Info("Failed to retrieve domain. Requeueing", "machineID", id)
-					r.queue.AddRateLimited(id)
-				} else {
-					log.Error(err, "failed to get machine state", "machineID", id)
-				}
-				continue
+	for {
+		select {
+		case ev, ok := <-eventChan:
+			if !ok {
+				log.Error(errors.New("missing lifecycle event"), "lifecycle event channel is closed")
```

Comment:
I would log the following error here: `Errorf("unexpected channel closure while waiting for data")`

Reply:
it isn't completely true because channel can be close when parent context is closed, what isn't unexpected. but no problem